### PR TITLE
Add extra check to callback return

### PR DESCRIPTION
**Pull request by thenotsoft**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #67

### CODE REVIEW

File: src/Exception/CallbackRuleException.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Yiisoft\Validator\Exception;
+
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+class CallbackRuleException extends \Exception implements FriendlyExceptionInterface
+{
+    public function getName(): string
+    {
+        return 'Invalid callable returned value';
```
Comment (viktorprogger) [inline src/Exception/CallbackRuleException.php:13]:
```suggestion
        return 'Invalid callable return value';
```

File: src/Exception/CallbackRuleException.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            The callback must return an instance of \Yiisoft\Validator\Result. An example of a valid callback:
+                static function (): \Yiisoft\Validator\Result
+                {
+                    \$result = new \Yiisoft\Validator\Result();
```
Comment (viktorprogger) [inline src/Exception/CallbackRuleException.php:22]:
```suggestion
                    \$result = new \\Yiisoft\\Validator\\Result();
```

File: src/Exception/CallbackRuleException.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public function getName(): string
+    {
+        return 'Invalid callable returned value';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            The callback must return an instance of \Yiisoft\Validator\Result. An example of a valid callback:
```
Comment (viktorprogger) [inline src/Exception/CallbackRuleException.php:19]:
```suggestion
            The callback must return an instance of \\Yiisoft\\Validator\\Result. An example of a valid callback:
```

File: src/Exception/CallbackRuleException.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Exception;
+
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+class CallbackRuleException extends \Exception implements FriendlyExceptionInterface
```
Comment (viktorprogger) [inline src/Exception/CallbackRuleException.php:9]: Let's simplify FQNS

File: src/Rule/Callback.php [hunk 5]
```diff
@@ -19,12 +23,21 @@ public function __construct(callable $callback)
...
         $callback = $this->callback;
-        /**
-         * @var $callbackResult Result
-         */
         $callbackResult = $callback($value, $dataSet);
+
+        if (!$callbackResult instanceof Result) {
+            throw new CallbackRuleException(
+                sprintf(
+                    'Return value of callback must be an instance of %s, %s returned.',
```
Comment (xepozz) [inline src/Rule/Callback.php:32]: Would be nice to encapsulate the exception message to class if you did create a new exception class

File: src/Rule/Callback.php [hunk 6]
```diff
@@ -19,12 +23,21 @@ public function __construct(callable $callback)
...
+                    Result::class,
+                    gettype($callbackResult)
```
Comment (xepozz) [inline src/Rule/Callback.php:34]: We need to add extra information about object class if `$callbackResult` is object but not implements `Result`
Else we will have the following messages:

> Return value of callback must be an instance of ..., **"object"** returned.

```suggestion
                    is_object($callbackResult) ? get_class($callbackResult) : gettype($callbackResult)
```

**Pull request merged**